File: src/validator/cs60_validator.py
import re

class CS60Validator:
  
  required_credit = {
    'วิชาแกน': 16, 
    'วิชาเฉพาะบังคับ': 55, 
    'วิชาเฉพาะเลือก': 21, 
    'ภาษากับการสื่อสาร': 13, 
    'พลเมืองไทยและพลเมืองโลก': 3, 
    'ศาสตร์แห่งผู้ประกอบการ': 3, 
    'สุนทรียศาสตร์': 3, 
    'อยู่ดีมีสุข': 3,
    'คณะวิทยาศาสตร์': 5,
  }
  
  def validate(self, student_data):
    # check credit
    course_categories = self.required_credit.keys()
    total_credit = student_data["total_credit_per_category"]
    
    for course_category in course_categories:
      if total_credit[course_category] < self.required_credit[course_category]:
        return False
      
    # check required course (e.g. P.E, Eng)
    enrolled_courses = student_data["enrolled_courses"]
    
    pe = 0
    eng = 0
    thai = 0
    knowledge_of_the_land = 0
    digital = 0
    science = 0
    
    for enrolled_course in enrolled_courses:
      course_id = enrolled_course["course_id"]
      
      # pe
      if re.search("01175\d{3}", course_id):
        pe += 1
            
      # eng
      if re.search("01355\d{3}", course_id):
        eng += 1
      
      # thai
      if re.search("01361101|01361102|01361103|01999021|01999022|02701011", course_id):
        thai += 1
      
      # knowledge of the land
      if re.search("01999111", course_id):
        knowledge_of_the_land += 1
      
      # digital
      if re.search("01418131|01420245|01420246", course_id):
        digital += 1
            
    if pe < 1 or eng < 3 or thai < 1 or knowledge_of_the_land < 1 or digital < 1:
      return False
    
    if sum([total_credit[c] for c in ['วิชาแกน', 'วิชาเฉพาะบังคับ', 'วิชาเฉพาะเลือก', 'ภาษากับการสื่อสาร', 'พลเมืองไทยและพลเมืองโลก', 'ศาสตร์แห่งผู้ประกอบการ']])
      
    # check grade
    if student_data["student_cum_gpa"] < 2:
      return False
      
    return True

File: internal/pkg/validator/cs60validator.go
package validator

import (
	"ku-transcript/internal/pkg/calculator"
	"ku-transcript/internal/pkg/utils/data"
	"ku-transcript/internal/pkg/utils/parser"
	"regexp"
)

func CS60Validator(student parser.Student) map[string]interface{} {
	requiredCredits := map[string]int{
		"วิชาแกน":                 16,
		"วิชาเฉพาะบังคับ":         55,
		"วิชาเฉพาะเลือก":          21,
		"ภาษากับการสื่อสาร":       13,
		"พลเมืองไทยและพลเมืองโลก": 3,
		"ศาสตร์แห่งผู้ประกอบการ":  3,
		"สุนทรียศาสตร์":           3,
		"อยู่ดีมีสุข":             3,
		"คณะวิทยาศาสตร์":          5,
	}
	minSumCredit := 128

	totalCredits, total := calculator.CalculateTotalCredits(data.ReadCourses("data"), student.EnrolledCourses, requiredCredits)
	isGraduated := true

	// Check if student has enough credits for each category
	for key, value := range totalCredits {
		if value < requiredCredits[key] {
			isGraduated = false
			break
		}
	}

	// Check if student enrolled in required courses
	pe := 0
	eng := 0
	thai := 0
	knowledgeOfTheLand := 0
	digital := 0

	peRegex := regexp.MustCompile(`01175\d{3}`)
	engRegex := regexp.MustCompile(`01355\d{3}`)
	thaiRegex := regexp.MustCompile(`01361101|01361102|01361103|01999021|01999022|02701011`)
	knowledgeOfTheLandRegex := regexp.MustCompile(`01999111`)
	digitalRegex := regexp.MustCompile(`01418131|01420245|01420246`)

	for _, course := range student.EnrolledCourses {
		courseID := course.CourseID

		if m := peRegex.MatchString(courseID); m {
			pe += 1
		}

		if m := engRegex.MatchString(courseID); m {
			eng += 1
		}

		if m := thaiRegex.MatchString(courseID); m {
			thai += 1
		}

		if m := knowledgeOfTheLandRegex.MatchString(courseID); m {
			knowledgeOfTheLand += 1
		}

		if m := digitalRegex.MatchString(courseID); m {
			digital += 1
		}
	}

	if pe < 1 || eng < 3 || thai < 1 || knowledgeOfTheLand < 1 || digital < 1 {
		isGraduated = false
	}

	// Check if student has enough total credits
	if total < minSumCredit {
		isGraduated = false
	}

	// Check cumulative GPA
	if student.StudentCumGPA < 2.00 {
		isGraduated = false
	}

	return map[string]interface{}{
		"total_credits": totalCredits,
		"total":         total,
		"is_graduated":  isGraduated,
	}
}
